### ISSUE TITLE
No method to bring itself online if it goes offline

### ISSUE THREAD

**Issue by jdc-cunningham** [jdc-cunningham/raspi-home-security-cam#3]
I started to work on this but couldn't get the bash scripts to self-execute or potentially a permission problem.

Lately the pi's been going down middle of the day.

I think I will also add an hourly poll to make sure it's still working (sends me a slack notification). Slack is currently my go to "mobile app notification" as I can't develop on mobile yet.

No errors either. I'm not sure if it's possibly a heat issue being that I stuffed all those components inside this container.

**Comment by jdc-cunningham** [comment]
Yeah today the PIR sensor/camera kept working but the WiFi connection went down.

**Comment by jdc-cunningham** [comment]
Alright... not going well so far. Do not edit /etc/sudoers without using visudo... hopefully I don't have to reflash this SD card and have to install everything... but I think I have a solution. Looks like since I was trying to use PHP (why) since I've used the exec command before to restart the pi when network is down. Had a permission problem. I will try to just bring up wifi (added ifconfig) to the /etc/sudoers file along with reboot/halt. Not sure if this is related actually now that I think about it but permissions sounds right, the cron is working/detecting that it's not connected bu the shell executed can't call shutdown or ifconfig.

**Comment by jdc-cunningham** [comment]
I used this stackexchange question about why PHP/Apache was not able to shutdown the pi.

Put the card in a Linux computer (Windows won't work) and can edit the /etc/sudoers file. (Only see boot in Windows). Will see now if this works.

https://raspberrypi.stackexchange.com/questions/26357/shutting-down-raspberry-pi-with-php

**Comment by jdc-cunningham** [comment]
Well I fixed the /etc/sudoers problem... but still can't get it to reboot
Will try from python approach, damn I'm tired

**Comment by jdc-cunningham** [comment]
hmmm

Tried this script
http://www.ridgesolutions.ie/index.php/2013/02/22/raspberry-pi-restart-shutdown-your-pi-from-python-code/

Problem now is it keeps shutting down/restarting (going on off network) ahhh

**Comment by jdc-cunningham** [comment]
Well... this sucks... not placing blame... hmm shoot might have to reflash this sd card, it's not the power supply, just swapped it out for a beefier 5V 3A adapter. It's rebooting so fast like I can't even SSH into it and run a command before I lose connection ahhhhh

**Comment by jdc-cunningham** [comment]
I was able to stop the restart loop. It occurred to me that the CRON schedule should allow the pi to reconnect so instead of doing every minute, do every 5 minutes so it has time to establish a wifi connection... I'll try it again with a 5 min delay just sucks waiting... or maybe 2 minutes is okay.

**Comment by jdc-cunningham** [comment]
Wow... it worked! Nice, the ridgesolutions script above is what I'm using. I didn't do the time poll part though... no it went down why... whyyyyy I'm not sure if this is an improper shutdown procedure or something.

Okay this will have to be on hold. I considered a dumb way to do this using a physical servo to flip a switch.

**Comment by jdc-cunningham** [comment]
Nope I'm dumb... it wasn't doing the if connected check so it always shuts down and restarts.

The script found here is much simpler.

https://www.raspberrypi.org/forums/viewtopic.php?t=133665

**Comment by jdc-cunningham** [comment]
Well at the very least there is a restart script. Set to an every 2 minutes cron schedule. Maybe overkill.

Did an IP check to one of my own domains, I don't know I feel bad polling other people's servers despite that the example.com one is probably designed to take it.

I don't know if I'll do the hourly poll for the pi to tell me that it's running... was trying to think of simple methods I could just drop into the cloudupload.py "module" for example where at the  top of the hour it would send out a slack notification.

At the very least there is a restart script. I'll add it to the repo.

I think the file-state (being on) will re-start the 3 threads when it restarts (after having lost connection in an on state) will have to test it out.

**Comment by jdc-cunningham** [comment]
Ahh man terrible name, but I've got it working. Had to add a state change to system-on.txt which is one of the checks that camera-check.py looks for. If the file is set to 'camera on' (testfile.txt changed by front end UI) 

Probably helps to have the order right.

So the system-restart.py (formerly wifi-reconnect.py) should run before the camera-check.py in the crontab.

Interesting... now I've got an asynchronous problem happening with CRON as the scripts aren't executed synchronously. I decided to just put the code from system-restart.py into the camera-check.py sucks it will poll every minute, but at least I'm targetting another device within my own network. This isn't great as if the wifi outside is down the system won't work. But that's another problem entirely I don't have a failsafe like a cellular data connection.

Oh man sorry Slack, dumb mistake kept trying to send notification over and over and over again.

Alright, satisfied with that, tested a couple of times with ```sudo ifconfig wlan0 down``` in a minute or so comes back online

edit: my "devkin" thing with the auto-timestamp would show here those sentences/paragraphs are several minutes apart.
